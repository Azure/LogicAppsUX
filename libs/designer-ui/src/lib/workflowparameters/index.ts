export * from './workflowparameter';
export * from './workflowparameters';
