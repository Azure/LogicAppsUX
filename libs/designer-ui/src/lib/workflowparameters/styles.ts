import { makeStyles, tokens } from '@fluentui/react-components';

export const useWorkflowParametersStyles = makeStyles({
  container: {
    display: 'flex',
    flexDirection: 'column',
    gap: tokens.spacingVerticalM,
  },

  heading: {
    marginTop: tokens.spacingVerticalXL,
    display: 'flex',
    gap: tokens.spacingHorizontalXS,
    flexWrap: 'wrap',
    alignItems: 'center',
    justifyContent: 'space-between',
  },

  addButton: {
    paddingLeft: '0px',
  },

  messageBar: {
    '& .fui-MessageBar': {
      padding: '8px 12px',
    },
  },

  listCell: {
    paddingBottom: tokens.spacingVerticalM,
    borderBottom: `1px solid ${tokens.colorNeutralStroke2}`,
    '&:last-child': {
      borderBottom: 'none',
    },
  },

  emptyState: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    marginTop: '8%',
    marginBottom: '30px',
  },

  emptyStateText: {
    textAlign: 'center',
    padding: '30px 50px',
  },

  parametersList: {
    display: 'flex',
    flexDirection: 'column',
  },

  parameterItem: {
    paddingBottom: tokens.spacingVerticalM,
    borderBottom: `1px solid ${tokens.colorNeutralStroke2}`,
    '&:last-child': {
      borderBottom: 'none',
    },
  },
});

export const useWorkflowParameterStyles = makeStyles({
  container: {
    display: 'flex',
    flexDirection: 'column',
    gap: tokens.spacingVerticalXS,
    padding: '16px 0px',
  },

  heading: {
    display: 'flex',
    gap: tokens.spacingHorizontalXS,
    alignItems: 'center',
  },

  headingButton: {
    flexGrow: 1,
  },

  typeBadge: {
    display: 'flex',
    alignItems: 'center',
    marginLeft: 'auto',
    marginRight: tokens.spacingHorizontalXS,
  },

  editOrDeleteButton: {
    display: 'flex',
    alignItems: 'center',
  },

  content: {
    padding: `0px ${tokens.spacingHorizontalM} ${tokens.spacingVerticalM} ${tokens.spacingHorizontalM}`,
    display: 'flex',
    flexDirection: 'column',
    gap: tokens.spacingVerticalS,
  },

  errorDot: {
    display: 'inline-block',
    marginLeft: tokens.spacingHorizontalXS,
  },

  field: {
    display: 'flex',
    flexDirection: 'column',
    gap: tokens.spacingVerticalXXS,
    minHeight: 'auto',
  },

  fieldLabel: {
    fontSize: tokens.fontSizeBase200,
    fontWeight: tokens.fontWeightSemibold,
    lineHeight: tokens.lineHeightBase200,
  },

  fieldEditor: {
    width: '100%', // Take full panel width
    '& .fui-Input': {
      width: '100%',
    },
    '& .fui-Dropdown': {
      width: '100%',
    },
  },

  valueField: {
    display: 'flex',
    flexDirection: 'column',
    gap: tokens.spacingVerticalXXS,
    alignItems: 'flex-start',
  },

  fieldError: {
    color: tokens.colorPaletteRedForeground1,
    fontSize: tokens.fontSizeBase200,
    marginTop: tokens.spacingVerticalXXS,
  },
});
