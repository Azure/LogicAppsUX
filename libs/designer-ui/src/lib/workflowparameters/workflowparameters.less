.msla-workflow-parameters {
  display: flex;
  flex-direction: column;
  gap: 16px;

  .msla-workflow-parameters-heading {
    margin-top: 20px;
    display: flex;
    gap: 8px;
    flex-wrap: wrap;
    align-items: center;
    justify-content: space-between;
  }

  .ms-List-cell {
    padding-bottom: 16px;
  }

  .msla-workflow-parameter {
    display: flex;
    gap: 8px;
    padding-bottom: 16px;
    border-bottom: 1px solid #9e9e9e;

    & > div:first-child {
      flex-grow: 1;
    }
  }

  .msla-workflow-parameters-empty {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin-top: 8%;
    margin-bottom: 30px;
  }

  .msla-workflow-parameters-error {
    display: flex;
    flex-direction: column;
  }

  .msla-theme-dark {
    .msla-workflow-parameters-text {
      color: @ms-color-themePrimary;
    }
  }

  .msla-workflow-parameter-read-only {
    vertical-align: sub;
  }

  .msla-workflow-parameter-heading-text {
    padding-left: 10px;
  }

  .msla-workflow-parameters-empty {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin-top: 8%;
    margin-bottom: 30px;
  }

  .msla-workflow-parameter-field {
    padding: 10px 0px 0px 10px;
    display: flex;
    min-height: 24px;
  }

  .msla-workflow-parameter-value-field {
    padding: 10px 0px 10px 10px;
    display: flex;
    align-items: flex-start;
  }
  .msla-workflow-parameters-text {
    text-align: center;
    padding: 30px 50px;
  }

  .msla-workflow-parameters-link {
    bottom: 20px;
  }

  .msla-workflow-parameters-link-text {
    padding-right: 5px;
  }

  .msla-workflow-parameter-error-dot {
    position: absolute;
    margin: -10px -4px;
  }

  *.ms-Button {
    height: 32px;
    font-size: 14px;
  }
}
