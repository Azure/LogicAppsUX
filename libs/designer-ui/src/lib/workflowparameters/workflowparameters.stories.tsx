import React from 'react';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import WorkflowParameters, { WorkflowParametersProps } from './workflowparameters';

export default {
  component: WorkflowParameters,
  title: 'Components/WorkflowParameter',
} as ComponentMeta<typeof WorkflowParameters>;
export const Standard: ComponentStory<typeof WorkflowParameters> = (args: WorkflowParametersProps) => <WorkflowParameters {...args} />;

Standard.args = {
  parameters: [
    {
      id: 'test1',
      defaultValue: 'true',
      type: 'Bool',
      name: 'test',
      isEditable: true,
    },
    {
      id: 'test2',
      defaultValue: '{}',
      type: 'Object',
      name: 'test2',
      isEditable: false,
    },
  ],
};
