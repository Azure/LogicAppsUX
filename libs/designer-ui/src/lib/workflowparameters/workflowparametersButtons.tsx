import type { EventHandler } from '../eventhandler';
import type { WorkflowParameterDefinition } from './workflowparameter';
import { Button, Tooltip } from '@fluentui/react-components';
import { bundleIcon, Delete24Filled, Delete24Regular, Edit24Filled, Edit24Regular } from '@fluentui/react-icons';
import { useIntl } from 'react-intl';

const DeleteIcon = bundleIcon(Delete24Filled, Delete24Regular);
const EditIcon = bundleIcon(Edit24Filled, Edit24Regular);

interface ButtonProps {
  onClick: React.MouseEventHandler<HTMLButtonElement>;
}

interface WorkflowParameterDeleteEvent {
  id: string;
}

type WorkflowParameterDeleteHandler = EventHandler<WorkflowParameterDeleteEvent>;

export interface EditOrDeleteButtonProps {
  showDelete?: boolean;
  onDelete?: WorkflowParameterDeleteHandler;
  definition: WorkflowParameterDefinition;
  setIsEditable: (value: boolean | ((prevVar: boolean | undefined) => boolean)) => void;
  setExpanded: (value: boolean | ((prevVar: boolean | undefined) => boolean)) => void;
}

export const EditOrDeleteButton = ({
  showDelete,
  onDelete,
  definition,
  setIsEditable,
  setExpanded,
}: EditOrDeleteButtonProps): JSX.Element => {
  const handleDelete = (e: React.MouseEvent<HTMLButtonElement>): void => {
    if (onDelete) {
      e.stopPropagation();
      onDelete({ id: definition.id });
    }
  };

  const handleEdit = (): void => {
    setIsEditable(true);
    setExpanded(true);
  };
  return showDelete ? <DeleteButton onClick={handleDelete} /> : <EditButton onClick={handleEdit} />;
};

function DeleteButton({ onClick }: ButtonProps): JSX.Element {
  const intl = useIntl();

  const deleteTitle = intl.formatMessage({
    defaultMessage: 'Delete Parameter',
    id: '5rkegy',
    description: 'Delete Button Tooltip Text',
  });

  return (
    <Tooltip relationship="label" content={deleteTitle}>
      <Button
        appearance="subtle"
        aria-label={deleteTitle}
        onClick={onClick}
        icon={<DeleteIcon style={{ color: 'var(--colorBrandForeground1)' }} />}
      />
    </Tooltip>
  );
}

function EditButton({ onClick }: ButtonProps): JSX.Element {
  const intl = useIntl();

  const editTitle = intl.formatMessage({
    defaultMessage: 'Edit parameter',
    id: '03RO5d',
    description: 'Edit Button Tooltip Text',
  });

  return (
    <Tooltip relationship="label" content={editTitle}>
      <Button
        appearance="subtle"
        data-testid="parameter-edit-icon-button"
        data-automation-id="parameter-edit-icon-button"
        aria-label={editTitle}
        onClick={onClick}
        icon={<EditIcon style={{ color: 'var(--colorBrandForeground1)' }} />}
      />
    </Tooltip>
  );
}
