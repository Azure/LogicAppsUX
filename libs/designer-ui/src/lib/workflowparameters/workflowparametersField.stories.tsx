import { ComponentMeta, ComponentStory } from '@storybook/react';
import { WorkflowparameterField, WorkflowparameterFieldProps } from './workflowparametersField';

export default {
  component: WorkflowparameterField,
  title: 'Components/WorkflowParameter',
} as ComponentMeta<typeof WorkflowparameterField>;
export const Field: ComponentStory<typeof WorkflowparameterField> = (args: WorkflowparameterFieldProps) => (
  <WorkflowparameterField {...args} />
);

Field.args = {
  isEditable: true,
  name: 'id',
  definition: { id: 'id', defaultValue: 'defaultValue', name: 'name', type: 'Array', value: 'testing' },
};
