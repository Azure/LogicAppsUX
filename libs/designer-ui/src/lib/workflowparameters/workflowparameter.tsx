import { CommandBarButton, IButtonStyles } from '@fluentui/react/lib/Button';
import { Dropdown, IDropdownOption, IDropdownStyles } from '@fluentui/react/lib/Dropdown';
import type { IIconProps } from '@fluentui/react/lib/Icon';
import { ILabelStyles, Label } from '@fluentui/react/lib/Label';
import { FontWeights, getTheme, IStyle } from '@fluentui/react/lib/Styling';
import { ITextStyles, Text } from '@fluentui/react/lib/Text';
import { ITextFieldProps, ITextFieldStyles, TextField } from '@fluentui/react/lib/TextField';
import React, { useState } from 'react';
import { equals, format } from '@microsoft-logic-apps/utils';
import Constants from '../constants';
import type { EventHandler } from '../eventhandler';
import { isHighContrastBlackOrInverted } from '../utils/theme';
import { useIntl } from 'react-intl';
import { EditOrDeleteButton } from './workflowparametersButtons';
import { ReadOnlyParameters } from './workflowparametersReadOnly';

const fieldStyles: IStyle = {
  display: 'inline-block',
  flexGrow: 1,
  flexShrink: 1,
  flexBasis: 'auto',
};

const disabledTextFieldStyles: Partial<ITextFieldStyles> = {
  root: fieldStyles,
  fieldGroup: {
    borderColor: Constants.FIELD_GROUP_BORDER_COLOR,
  },
};

const dropdownStyles: Partial<IDropdownStyles> = {
  root: fieldStyles,
};

const commandBarStyles: Partial<IButtonStyles> = {
  label: {
    fontWeight: FontWeights.semibold,
  },
};

export const labelStyles: Partial<ILabelStyles> = {
  root: {
    display: 'inline-block',
    minWidth: 100,
    verticalAlign: 'top',
  },
};

const textStyles: Partial<ITextStyles> = {
  root: {
    color: getTheme().palette.yellow,
    fontWeight: FontWeights.bold,
  },
};

const textFieldStyles: Partial<ITextFieldStyles> = {
  root: fieldStyles,
};

const textFieldWithWarningStyles: Partial<ITextFieldStyles> = {
  root: fieldStyles,
  fieldGroup: {
    borderColor: Constants.FIELD_GROUP_BORDER_COLOR_WARNING,
    selectors: {
      '&:hover': {
        borderColor: Constants.FIELD_GROUP_BORDER_COLOR_WARNING,
      },
    },
  },
};

const NAME_KEY = 'name';
const DEFAULT_VALUE_KEY = 'defaultValue';

export interface WorkflowParameterUpdateEvent {
  id: string;
  newDefinition: WorkflowParameterDefinition;
}

export interface WorkflowParameterDeleteEvent {
  id: string;
}

export type WorkflowParameterUpdateHandler = EventHandler<WorkflowParameterUpdateEvent>;
export type WorkflowParameterDeleteHandler = EventHandler<WorkflowParameterDeleteEvent>;
export type RegisterLanguageHandler = () => void;

export interface WorkflowParameterDefinition {
  defaultValue?: string;
  id: string;
  isEditable?: boolean;
  name?: string;
  type?: string;
  value?: string;
}

export interface WorkflowParameterProps {
  definition: WorkflowParameterDefinition;
  isReadOnly?: boolean;
  validationErrors?: Record<string, string>;
  onChange?: WorkflowParameterUpdateHandler;
  onDelete?: WorkflowParameterDeleteHandler;
  onRegisterLanguageProvider?: RegisterLanguageHandler;
}

export interface WorkflowParameterState {
  defaultValue?: string;
  expanded?: boolean;
  isEditable?: boolean;
  isInverted?: boolean;
  name?: string;
  type?: string;
  valueWarningMessage?: string | undefined;
}

export interface ParameterFieldDetails {
  name: string;
  defaultValue: string;
  type: string;
  value: string;
}

export function WorkflowParameter({ definition, validationErrors, isReadOnly, onChange, ...props }: WorkflowParameterProps): JSX.Element {
  const [defaultValue, setDefaultValue] = useState(definition.defaultValue);
  const [expanded, setExpanded] = useState(!!definition.isEditable);
  const [isEditable, setIsEditable] = useState(definition.isEditable);
  const [isInverted, setIsInverted] = useState(isHighContrastBlackOrInverted());
  const [name, setName] = useState(definition.name);
  const [type, setType] = useState(definition.type);
  const [valueWarningMessage, setValueWarningMessage] = useState(getValueWarningMessage(definition.defaultValue, definition.type));
  const intl = useIntl();

  const typeOptions: IDropdownOption[] = [
    {
      key: Constants.WORKFLOW_PARAMETER_SERIALIZED_TYPE.ARRAY,
      text: intl.formatMessage({
        defaultMessage: 'Array',
        description: 'This is an option in a dropdown where users can select type Array for their parameter.',
      }),
    },
    {
      key: Constants.WORKFLOW_PARAMETER_SERIALIZED_TYPE.BOOL,
      text: intl.formatMessage({
        defaultMessage: 'Bool',
        description: 'This is an option in a dropdown where users can select type Boolean for their parameter.',
      }),
    },
    {
      key: Constants.WORKFLOW_PARAMETER_SERIALIZED_TYPE.FLOAT,
      text: intl.formatMessage({
        defaultMessage: 'Float',
        description: 'This is an option in a dropdown where users can select type Float for their parameter.',
      }),
    },
    {
      key: Constants.WORKFLOW_PARAMETER_SERIALIZED_TYPE.INT,
      text: intl.formatMessage({
        defaultMessage: 'Int',
        description: 'This is an option in a dropdown where users can select type Integer for their parameter.',
      }),
    },
    {
      key: Constants.WORKFLOW_PARAMETER_SERIALIZED_TYPE.OBJECT,
      text: intl.formatMessage({
        defaultMessage: 'Object',
        description: 'This is an option in a dropdown where users can select type Object for their parameter.',
      }),
    },
    {
      key: Constants.WORKFLOW_PARAMETER_SERIALIZED_TYPE.STRING,
      text: intl.formatMessage({
        defaultMessage: 'String',
        description: 'This is an option in a dropdown where users can select type String for their parameter.',
      }),
    },
  ];

  const parameterDetails: ParameterFieldDetails = {
    name: `${definition.id}-${NAME_KEY}`,
    defaultValue: `${definition.id}-${DEFAULT_VALUE_KEY}`,
    type: `${definition.id}-type`,
    value: `${definition.id}-value`,
  };
  const errors = validationErrors ? validationErrors : {};
  const iconProps: IIconProps = {
    iconName: expanded ? 'ChevronDownMed' : 'ChevronRightMed',
    styles: {
      root: {
        fontSize: 14,
        color: isInverted ? 'white' : '#514f4e',
      },
    },
  };

  const handleToggleExpand = (): void => {
    setExpanded(!expanded);
  };

  const renderParameterFields = (parameterDetails: ParameterFieldDetails, errors: Record<string, string>): JSX.Element => {
    const nameTitle = intl.formatMessage({
      defaultMessage: 'Name',
      description: 'Parameter Field Name Title',
    });
    const nameDescription = intl.formatMessage({
      defaultMessage: 'Enter parameter name.',
      description: 'Parameter Field Name Description',
    });
    const typeTitle = intl.formatMessage({
      defaultMessage: 'Type',
      description: 'Parameter Field Type Title',
    });
    if (isEditable) {
      return (
        <>
          <div className="msla-workflow-parameter-field">
            <Label styles={labelStyles} required={true} htmlFor={parameterDetails.name}>
              {nameTitle}
            </Label>
            <TextField
              styles={textFieldStyles}
              id={parameterDetails.name}
              ariaLabel={nameTitle}
              placeholder={nameDescription}
              value={name}
              errorMessage={errors[NAME_KEY]}
              onChange={onNameChange}
              disabled={isReadOnly}
            />
          </div>
          <div className="msla-workflow-parameter-field">
            <Label styles={labelStyles} required={true} htmlFor={parameterDetails.type}>
              {typeTitle}
            </Label>
            <Dropdown
              id={parameterDetails.type}
              ariaLabel={typeTitle}
              options={typeOptions}
              selectedKey={type}
              styles={dropdownStyles}
              onChange={onTypeChange}
              disabled={isReadOnly}
            />
          </div>
          {renderParameterValueField(parameterDetails, errors)}
        </>
      );
    } else {
      return (
        <ReadOnlyParameters
          name={name}
          defaultValue={defaultValue}
          type={type}
          definition={definition}
          parameterDetails={parameterDetails}
        />
      );
    }
  };

  const onNameChange = (_event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string): void => {
    setName(newValue);
    if (onChange) {
      onChange({
        id: definition.id,
        newDefinition: { ...definition, name: newValue },
      });
    }
  };

  const onTypeChange = (_event?: React.FormEvent<HTMLDivElement>, item?: IDropdownOption): void => {
    const newType = item?.key.toString();

    if (onChange) {
      onChange({
        id: definition.id,
        newDefinition: { ...definition, type },
      });
    }
    setType(newType);
    setValueWarningMessage(getValueWarningMessage(defaultValue, newType));
  };

  const renderParameterValueField = (parameterDetails: ParameterFieldDetails, errors: Record<string, string>): JSX.Element => {
    const defaultValueTitle = intl.formatMessage({
      defaultMessage: 'Default Value',
      description: 'Parameter Field Default Value Title',
    });
    const defaultValueDescription = intl.formatMessage({
      defaultMessage: 'Enter default value for parameter.',
      description: 'Parameter Field Default Value Placeholder Text',
    });
    const actualValueTitle = intl.formatMessage({
      defaultMessage: 'Actual Value',
      description: 'Parameter Field Actual Value Title',
    });
    return (
      <>
        <div className="msla-workflow-parameter-field">
          <Label styles={labelStyles} htmlFor={parameterDetails.defaultValue}>
            {defaultValueTitle}
          </Label>
          <TextField
            id={parameterDetails.defaultValue}
            ariaLabel={defaultValueTitle}
            placeholder={defaultValueDescription}
            description={valueWarningMessage}
            value={defaultValue}
            errorMessage={errors[DEFAULT_VALUE_KEY]}
            styles={valueWarningMessage ? textFieldWithWarningStyles : textFieldStyles}
            onChange={onValueChange}
            onRenderDescription={valueWarningMessage ? onRenderDescription : undefined}
            disabled={isReadOnly}
          />
        </div>
        <div className="msla-workflow-parameter-field">
          <Label styles={labelStyles} htmlFor={parameterDetails.value}>
            {actualValueTitle}
          </Label>
          <TextField
            styles={disabledTextFieldStyles}
            id={parameterDetails.value}
            ariaLabel={actualValueTitle}
            type={isSecureParameter(type) ? 'password' : undefined}
            defaultValue={definition.value}
            disabled
          />
        </div>
      </>
    );
  };

  const onValueChange = (_event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string): void => {
    handleContentChange(newValue);
  };

  const handleContentChange = (value?: string) => {
    setDefaultValue(value);
    setValueWarningMessage(getValueWarningMessage(value, type));

    if (onChange) {
      onChange({
        id: definition.id,
        newDefinition: { ...definition, defaultValue: value },
      });
    }
  };

  const onRenderDescription = (props?: ITextFieldProps): JSX.Element => {
    return (
      <Text variant="small" styles={textStyles}>
        {props?.description}
      </Text>
    );
  };

  const headingTitle = intl.formatMessage({
    defaultMessage: 'New parameter',
    description: 'Heading Title for a Parameter Without Name',
  });

  return (
    <div className="msla-workflow-parameter">
      <div className="msla-workflow-parameter-group">
        <div>
          <CommandBarButton
            className="msla-workflow-parameter-heading-button"
            iconProps={iconProps}
            onClick={handleToggleExpand}
            styles={commandBarStyles}
            text={name ? name : headingTitle}
          />
        </div>
        {expanded ? renderParameterFields(parameterDetails, errors) : null}
      </div>
      {!isReadOnly ? (
        <div className="msla-workflow-parameter-edit-or-delete-button">
          <EditOrDeleteButton
            onDelete={props.onDelete}
            showDelete={isEditable}
            definition={definition}
            setIsEditable={setIsEditable}
            setExpanded={setExpanded}
          />
        </div>
      ) : null}
    </div>
  );
}

function isSecureParameter(type?: string): boolean {
  return equals(type, Constants.WORKFLOW_PARAMETER_TYPE.SECURE_STRING) || equals(type, Constants.WORKFLOW_PARAMETER_TYPE.SECURE_OBJECT);
}

function getValueWarningMessage(value?: string, type?: string): string | undefined {
  return isSecureParameter(type) && !!value ? format('Warning Message', type) : undefined;
}
