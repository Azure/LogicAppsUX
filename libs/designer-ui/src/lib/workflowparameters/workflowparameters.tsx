import ParametersIcon from '../card/images/parameters.svg';
import Constants from '../constants';
import { isHighContrastBlack } from '../utils/theme';
import type { WorkflowParameterDefinition, WorkflowParameterDeleteHandler, WorkflowParameterUpdateHandler } from './workflowparameter';
import { WorkflowParameter } from './workflowparameter';
import { IconButton, CommandBarButton, Icon, Link, List, MessageBar, useTheme, Text } from '@fluentui/react';
import type { IIconProps, IIconStyles, IMessageBarStyles } from '@fluentui/react';
import { useIntl } from 'react-intl';

const navigateIconStyle: IIconStyles = {
  root: {
    color: Constants.BRAND_COLOR,
  },
};

const lightMessageBarStyles: IMessageBarStyles = {
  root: {
    backgroundColor: Constants.BACKGROUND_COLOR_LIGHT,
  },
  icon: {
    color: 'gray',
  },
  text: {
    color: Constants.BACKGROUND_COLOR_DARK,
  },
};

const darkMessageBarStyles: IMessageBarStyles = {
  root: {
    backgroundColor: Constants.BACKGROUND_COLOR_DARK,
  },
  icon: {
    color: 'white',
  },
  text: {
    color: 'white',
  },
};

const InfoBar = ({ isInverted }: { isInverted: boolean }) => {
  const intl = useIntl();
  const text = intl.formatMessage({
    defaultMessage:
      'The parameters will be saved when the workflow is saved. You can edit it here before save or edit it in the parameter page after save.',
    description: 'Text for Info Bar',
  });
  return (
    <div className="msla-workflow-parameters-message-bar">
      <MessageBar isMultiline={true} styles={isInverted ? darkMessageBarStyles : lightMessageBarStyles}>
        {text}
      </MessageBar>
    </div>
  );
};

type OnClickHandler = () => void;

export interface WorkflowParametersProps {
  isEditable?: boolean;
  isReadOnly?: boolean;
  useLegacy?: boolean;
  parameters: WorkflowParameterDefinition[];
  validationErrors?: Record<string, Record<string, string | undefined>>;
  onDismiss?: OnClickHandler;
  onAddParameter?: OnClickHandler;
  onDeleteParameter?: WorkflowParameterDeleteHandler;
  onManageParameters?: OnClickHandler;
  onRegisterLanguageProvider?: OnClickHandler;
  onUpdateParameter?: WorkflowParameterUpdateHandler;
}

export interface WorkflowParametersState {
  isInverted: boolean;
}

export function WorkflowParameters({
  parameters = [],
  isReadOnly,
  useLegacy,
  onDismiss,
  onManageParameters,
  onAddParameter,
  onDeleteParameter,
  onUpdateParameter,
  validationErrors,
  onRegisterLanguageProvider,
}: WorkflowParametersProps): JSX.Element {
  const theme = useTheme();
  const isInverted = isHighContrastBlack() || theme.isInverted;

  const intl = useIntl();

  const addIcon: IIconProps = { iconName: 'Add' };

  const handleAddParameter = (event: React.MouseEvent<HTMLButtonElement>): void => {
    if (onAddParameter) {
      event.stopPropagation();
      onAddParameter();
    }
  };

  const renderTitleAndDescription = (): JSX.Element => {
    const description1 = useLegacy
      ? intl.formatMessage({
          defaultMessage: 'Create, manage Logic Apps parameters, give it a default value.',
          description: 'Description for Workflow Parameters Part 1 for Legacy Parameters mode.',
        })
      : intl.formatMessage({
          defaultMessage: 'Parameters are shared across workflows in a Logic App.',
          description: 'Description for Workflow Parameters Part 1',
        });
    const description2 = useLegacy
      ? intl.formatMessage({
          defaultMessage:
            'Parameters used in Logic App will be converted into Azure Resource Manager template during deployment template generation.',
          description: 'Description for Workflow Parameters Part 2 for Legacy Parameters mode.',
        })
      : intl.formatMessage({
          defaultMessage: 'To reference a parameter, use the dynamic content list.',
          description: 'Description for Workflow Parameters Part 2',
        });

    return (
      <div className="msla-workflow-parameters-empty">
        <img src={ParametersIcon} alt="" role="presentation" />
        <div className="msla-workflow-parameters-text">
          <p>{description1}</p>
          <p>{description2}</p>
        </div>
      </div>
    );
  };

  const ManageParametersLink = (): JSX.Element => {
    const intl = useIntl();

    const jsonText = intl.formatMessage({
      defaultMessage: 'Edit in JSON',
      description: 'Parameter Link Text',
    });
    return (
      <footer className="msla-workflow-parameters-link">
        <Link className="msla-workflow-parameters-link-text" onClick={onManageParameters}>
          {jsonText}
        </Link>
        <Icon iconName="NavigateExternalInline" styles={navigateIconStyle} className="msla-workflow-parameters-link-icon" />
      </footer>
    );
  };

  const renderParameter = (item?: WorkflowParameterDefinition): JSX.Element => {
    // TODO: 12798972 Workflow Parameter
    const parameterErrors = validationErrors && item ? validationErrors[item.id] : undefined;
    return (
      <WorkflowParameter
        key={item?.id}
        definition={item ?? { id: 'id', type: Constants.WORKFLOW_PARAMETER_SERIALIZED_TYPE.ARRAY }}
        isReadOnly={isReadOnly}
        useLegacy={useLegacy}
        isInverted={isInverted}
        onChange={onUpdateParameter}
        onDelete={onDeleteParameter}
        onRegisterLanguageProvider={onRegisterLanguageProvider}
        validationErrors={parameterErrors}
      />
    );
  };

  const createText = intl.formatMessage({
    defaultMessage: 'Create parameter',
    description: 'Create Parameter Text',
  });
  const titleText = intl.formatMessage({
    defaultMessage: 'Parameters',
    description: 'Workflow Parameters Title',
  });
  const onClose = () => onDismiss?.();

  return (
    <div className="msla-workflow-parameters">
      <div className="msla-workflow-parameters-heading">
        <Text variant="xLarge">{titleText}</Text>
        <IconButton onClick={onClose} iconProps={{ iconName: 'Cancel' }} />
      </div>

      {useLegacy ? null : <InfoBar isInverted={isInverted} />}
      <div className="msla-workflow-parameters-add">
        <CommandBarButton disabled={isReadOnly} text={createText} iconProps={addIcon} onClick={handleAddParameter} />
      </div>
      {parameters.length ? null : renderTitleAndDescription()}
      {parameters.length ? <List items={parameters} onRenderCell={renderParameter} /> : null}
      {onManageParameters ? <ManageParametersLink /> : null}
    </div>
  );
}
