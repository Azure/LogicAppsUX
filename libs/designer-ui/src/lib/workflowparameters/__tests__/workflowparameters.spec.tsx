import * as React from 'react';
import * as ReactShallowRenderer from 'react-test-renderer/shallow';
import { WorkflowParameters, WorkflowParametersProps } from '../workflowparameters';
import { initializeIcons } from '@fluentui/react';
import { useIntl } from 'react-intl';

describe('ui/workflowparameters/workflowparameters', () => {
  let minimal: WorkflowParametersProps, renderer: ReactShallowRenderer.ShallowRenderer;

  beforeEach(() => {
    minimal = {
      parameters: [
        {
          id: 'test1',
          defaultValue: 'true',
          type: 'Bool',
          name: 'test',
          isEditable: true,
        },
        {
          id: 'test2',
          defaultValue: '{}',
          type: 'Object',
          name: 'test2',
          isEditable: false,
        },
      ],
    };
    renderer = ReactShallowRenderer.createRenderer();
    initializeIcons();
  });

  afterEach(() => {
    renderer.unmount();
  });

  it('should construct.', () => {
    const parameters = renderer.render(<WorkflowParameters {...minimal} />);
    expect(parameters).toMatchSnapshot();
  });

  it('should render parameters when provided.', () => {
    const intl = useIntl();
    renderer.render(<WorkflowParameters {...minimal} />);
    const parameters = renderer.getRenderOutput();
    const [header, messageBar, add, parameterList]: any[] = React.Children.toArray(parameters.props.children);

    const headerTitle = intl.formatMessage({
      defaultMessage: 'Parameters',
      description: 'Create Title',
    });
    expect(header.props.className).toBe('msla-workflow-parameters-create');
    const headerText = header.props.children;
    expect(headerText.props.defaultMessage[0].value).toBe(headerTitle);

    expect(messageBar).toBeDefined();

    const addMessage = intl.formatMessage({
      defaultMessage: 'Create parameter',
      description: 'Create Parameter Text',
    });
    expect(add.props.className).toBe('msla-workflow-parameters-add');
    const addButton = add.props.children;
    expect(addButton.props.text).toBe(addMessage);
    expect(addButton.props.className).toBe('msla-workflow-parameters-create-button');

    expect(parameterList.props.items).toHaveLength(2);
  });
});
