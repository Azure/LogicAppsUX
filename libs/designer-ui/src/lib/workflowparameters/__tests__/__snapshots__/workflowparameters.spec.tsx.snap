// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/workflowparameters/workflowparameters should construct. 1`] = `
<div
  className="msla-workflow-parameters"
>
  <div
    className="msla-workflow-parameters-heading"
  >
    <Text
      variant="xLarge"
    >
      Parameters
    </Text>
    <Button
      appearance="subtle"
      icon={<CompoundIcon />}
      onClick={[Function]}
    />
  </div>
  <InfoBar />
  <div
    className="msla-workflow-parameters-add"
  >
    <Button
      icon={<CompoundIcon />}
      onClick={[Function]}
    >
      Create parameter
    </Button>
  </div>
  <List
    items={
      Array [
        Object {
          "id": "test1",
          "isEditable": true,
          "name": "test",
          "type": "Bool",
          "value": "true",
        },
        Object {
          "id": "test2",
          "isEditable": false,
          "name": "test2",
          "type": "Object",
          "value": "{}",
        },
      ]
    }
    onRenderCell={[Function]}
    renderedWindowsAhead={2}
    renderedWindowsBehind={2}
    startIndex={0}
  />
</div>
`;
